File: validation/skmisc/validate.py
import numpy as np
from skmisc.loess import loess
import json
import os

OUTPUT_DIR = "output/scikit/"
os.makedirs(OUTPUT_DIR, exist_ok=True)

def run_scenario(name, x, y, frac, deg, iter, notes="", **kwargs):
    print(f"Running scenario: {name}")
    
    # Configure LOESS
    control = {'iterations': iter}
    l = loess(x, y, span=frac, degree=deg, control=control, **kwargs)
    l.fit()
    
    fitted = l.predict(x, stderror=False).values
    
    # Sanitize for JSON (numpy types)
    x_list = x.tolist()
    y_list = y.tolist()
    fitted_list = fitted.tolist()
    
    data = {
        "name": name,
        "notes": notes,
        "input": {
            "x": x_list,
            "y": y_list
        },
        "params": {
            "fraction": frac,
            "degree": deg,
            "iterations": iter,
            "extra": kwargs
        },
        "result": {
            "fitted": fitted_list
        }
    }
    
    path = os.path.join(OUTPUT_DIR, f"{name}.json")
    with open(path, 'w') as f:
        json.dump(data, f, indent=2)

def generate_data(n=100, kind='linear', noise=0.0, range_min=0.0, range_max=1.0, outlier_ratio=0.0):
    np.random.seed(42) # Fixed seed for reproducibility
    x = np.linspace(range_min, range_max, n)
    
    if kind == 'linear':
        y = 2 * x + 1
    elif kind == 'quadratic':
        y = x**2
    elif kind == 'sine':
        y = np.sin(4 * x)
    elif kind == 'step':
        y = np.where(x < (range_min + range_max)/2, 0.0, 1.0)
    elif kind == 'constant':
        y = np.full_like(x, 5.0)
    else:
        y = x.copy()
        
    # Add noise
    if noise > 0:
        y += np.random.normal(0, noise, n)
        
    # Add outliers
    if outlier_ratio > 0:
        n_out = int(n * outlier_ratio)
        indices = np.random.choice(n, n_out, replace=False)
        y[indices] += 10.0 # Significant outlier
        
    return x, y

def main():
    # 1. Tiny Linear
    x, y = generate_data(n=10, kind='linear')
    run_scenario("01_tiny_linear", x, y, frac=0.8, deg=1, iter=0)
    
    # 2. Small Quadratic
    x, y = generate_data(n=50, kind='quadratic')
    run_scenario("02_small_quadratic", x, y, frac=0.5, deg=2, iter=0)
    
    # 3. Sine Standard
    x, y = generate_data(n=100, kind='sine', noise=0.1)
    run_scenario("03_sine_standard", x, y, frac=0.3, deg=1, iter=0)
    
    # 4. Sine Robust
    x, y = generate_data(n=100, kind='sine', outlier_ratio=0.05)
    run_scenario("04_sine_robust", x, y, frac=0.3, deg=1, iter=4)
    
    # 5. Degree 0
    x, y = generate_data(n=100, kind='sine')
    run_scenario("05_degree_0", x, y, frac=0.2, deg=0, iter=0)
    
    # 6. Large scale
    x, y = generate_data(n=500, kind='sine')
    run_scenario("06_large_scale", x, y, frac=0.1, deg=1, iter=0)
    
    # 7. High Smoothness
    x, y = generate_data(n=100, kind='linear', noise=0.5)
    run_scenario("07_high_smoothness", x, y, frac=0.9, deg=1, iter=0)
    
    # 8. Low Smoothness
    x, y = generate_data(n=100, kind='sine')
    run_scenario("08_low_smoothness", x, y, frac=0.05, deg=1, iter=0, surface='direct')
    
    # 9. Quadratic Robust
    x, y = generate_data(n=100, kind='quadratic', outlier_ratio=0.1)
    run_scenario("09_quadratic_robust", x, y, frac=0.5, deg=2, iter=4)
    
    # 10. Constant Function
    x, y = generate_data(n=50, kind='constant')
    run_scenario("10_constant", x, y, frac=0.5, deg=1, iter=0)
    
    # 11. Step Function
    x, y = generate_data(n=100, kind='step')
    run_scenario("11_step_func", x, y, frac=0.4, deg=1, iter=0)
    
    # 12. End-effects Left
    x, y = generate_data(n=50, kind='linear', noise=0.1)
    run_scenario("12_end_effects_left", x, y, frac=0.3, deg=1, iter=0, notes="Check left boundary")
    
    # 13. End-effects Right (same data, just naming)
    run_scenario("13_end_effects_right", x, y, frac=0.3, deg=1, iter=0, notes="Check right boundary")
    
    # 14. Sparse Data
    x, y = generate_data(n=20, range_max=100.0, kind='linear', noise=1.0)
    run_scenario("14_sparse_data", x, y, frac=0.6, deg=1, iter=0)
    
    # 15. Dense Data
    x, y = generate_data(n=1000, kind='sine', noise=0.1)
    run_scenario("15_dense_data", x, y, frac=0.01, deg=1, iter=0, surface='direct')
    
    # 16. Degree 2 Sine
    x, y = generate_data(n=100, kind='sine')
    run_scenario("16_degree_2_sine", x, y, frac=0.4, deg=2, iter=0)
    
    # 17. Robust Degree 0
    x, y = generate_data(n=100, kind='linear', outlier_ratio=0.05)
    run_scenario("17_robust_degree_0", x, y, frac=0.4, deg=0, iter=4)
    
    # 18. Iter 2 Check
    x, y = generate_data(n=100, kind='sine', outlier_ratio=0.05)
    run_scenario("18_iter_2", x, y, frac=0.4, deg=1, iter=2)
    
    # 19. Interpolate Exact
    x, y = generate_data(n=50, kind='linear')
    run_scenario("19_interpolate_exact", x, y, frac=0.5, deg=1, iter=0)
    
    # 20. Zero Variance
    x, y = generate_data(n=10, kind='constant') # all 5.0
    run_scenario("20_zero_variance", x, y, frac=0.5, deg=1, iter=0)

if __name__ == "__main__":
    main()

File: validation/R/validate.R
#!/usr/bin/env Rscript
# R LOESS Validation Script
# Generates reference outputs for all validation scenarios

library(jsonlite)

OUTPUT_DIR <- "output/r/"
dir.create(OUTPUT_DIR, showWarnings = FALSE, recursive = TRUE)

run_scenario <- function(name, x, y, frac, deg, iter, notes = "", ...) {
  cat(sprintf("Running scenario: %s\n", name))

  # Configure LOESS
  # In R: family="symmetric" enables robustness, family="gaussian" disables it
  if (iter > 0) {
    # Robustness enabled
    fit <- loess(y ~ x,
      span = frac, degree = deg,
      family = "symmetric", iterations = iter, ...
    )
  } else {
    # No robustness (family="gaussian" doesn't use iterations parameter)
    fit <- loess(y ~ x,
      span = frac, degree = deg,
      family = "gaussian", ...
    )
  }
  fitted <- fitted(fit)

  # Create output structure
  data <- list(
    name = name,
    notes = notes,
    input = list(
      x = x,
      y = y
    ),
    params = list(
      fraction = frac,
      degree = deg,
      iterations = iter,
      family = if (iter > 0) "symmetric" else "gaussian",
      extra = list(...)
    ),
    result = list(
      fitted = fitted
    )
  )

  # Save to JSON
  path <- file.path(OUTPUT_DIR, paste0(name, ".json"))
  write_json(data, path, auto_unbox = TRUE, pretty = TRUE, digits = NA)
}

generate_data <- function(n = 100, kind = "linear", noise = 0.0,
                          range_min = 0.0, range_max = 1.0, outlier_ratio = 0.0) {
  # Fixed seed for reproducibility (same as Python)
  set.seed(42)

  x <- seq(range_min, range_max, length.out = n)

  if (kind == "linear") {
    y <- 2 * x + 1
  } else if (kind == "quadratic") {
    y <- x^2
  } else if (kind == "sine") {
    y <- sin(4 * x)
  } else if (kind == "step") {
    y <- ifelse(x < (range_min + range_max) / 2, 0.0, 1.0)
  } else if (kind == "constant") {
    y <- rep(5.0, n)
  } else {
    y <- x
  }

  # Add noise
  if (noise > 0) {
    y <- y + rnorm(n, 0, noise)
  }

  # Add outliers
  if (outlier_ratio > 0) {
    n_out <- as.integer(n * outlier_ratio)
    indices <- sample(n, n_out, replace = FALSE)
    y[indices] <- y[indices] + 10.0 # Significant outlier
  }

  list(x = x, y = y)
}

main <- function() {
  # 1. Tiny Linear
  data <- generate_data(n = 10, kind = "linear")
  run_scenario("01_tiny_linear", data$x, data$y, frac = 0.8, deg = 1, iter = 0)

  # 2. Small Quadratic
  data <- generate_data(n = 50, kind = "quadratic")
  run_scenario("02_small_quadratic", data$x, data$y, frac = 0.5, deg = 2, iter = 0)

  # 3. Sine Standard
  data <- generate_data(n = 100, kind = "sine", noise = 0.1)
  run_scenario("03_sine_standard", data$x, data$y, frac = 0.3, deg = 1, iter = 0)

  # 4. Sine Robust
  data <- generate_data(n = 100, kind = "sine", outlier_ratio = 0.05)
  run_scenario("04_sine_robust", data$x, data$y, frac = 0.3, deg = 1, iter = 4)

  # 5. Degree 0
  data <- generate_data(n = 100, kind = "sine")
  run_scenario("05_degree_0", data$x, data$y, frac = 0.2, deg = 0, iter = 0)

  # 6. Large scale
  data <- generate_data(n = 500, kind = "sine")
  run_scenario("06_large_scale", data$x, data$y, frac = 0.1, deg = 1, iter = 0)

  # 7. High Smoothness
  data <- generate_data(n = 100, kind = "linear", noise = 0.5)
  run_scenario("07_high_smoothness", data$x, data$y, frac = 0.9, deg = 1, iter = 0)

  # 8. Low Smoothness
  data <- generate_data(n = 100, kind = "sine")
  run_scenario("08_low_smoothness", data$x, data$y,
    frac = 0.05, deg = 1, iter = 0,
    surface = "direct"
  )

  # 9. Quadratic Robust
  data <- generate_data(n = 100, kind = "quadratic", outlier_ratio = 0.1)
  run_scenario("09_quadratic_robust", data$x, data$y, frac = 0.5, deg = 2, iter = 4)

  # 10. Constant Function
  data <- generate_data(n = 50, kind = "constant")
  run_scenario("10_constant", data$x, data$y, frac = 0.5, deg = 1, iter = 0)

  # 11. Step Function
  data <- generate_data(n = 100, kind = "step")
  run_scenario("11_step_func", data$x, data$y, frac = 0.4, deg = 1, iter = 0)

  # 12. End-effects Left
  data <- generate_data(n = 50, kind = "linear", noise = 0.1)
  run_scenario("12_end_effects_left", data$x, data$y,
    frac = 0.3, deg = 1, iter = 0,
    notes = "Check left boundary"
  )

  # 13. End-effects Right (same data, just naming)
  run_scenario("13_end_effects_right", data$x, data$y,
    frac = 0.3, deg = 1, iter = 0,
    notes = "Check right boundary"
  )

  # 14. Sparse Data
  data <- generate_data(n = 20, range_max = 100.0, kind = "linear", noise = 1.0)
  run_scenario("14_sparse_data", data$x, data$y, frac = 0.6, deg = 1, iter = 0)

  # 15. Dense Data
  data <- generate_data(n = 1000, kind = "sine", noise = 0.1)
  run_scenario("15_dense_data", data$x, data$y,
    frac = 0.01, deg = 1, iter = 0,
    surface = "direct"
  )

  # 16. Degree 2 Sine
  data <- generate_data(n = 100, kind = "sine")
  run_scenario("16_degree_2_sine", data$x, data$y, frac = 0.4, deg = 2, iter = 0)

  # 17. Robust Degree 0
  data <- generate_data(n = 100, kind = "linear", outlier_ratio = 0.05)
  run_scenario("17_robust_degree_0", data$x, data$y, frac = 0.4, deg = 0, iter = 4)

  # 18. Iter 2 Check
  data <- generate_data(n = 100, kind = "sine", outlier_ratio = 0.05)
  run_scenario("18_iter_2", data$x, data$y, frac = 0.4, deg = 1, iter = 2)

  # 19. Interpolate Exact
  data <- generate_data(n = 50, kind = "linear")
  run_scenario("19_interpolate_exact", data$x, data$y, frac = 0.5, deg = 1, iter = 0)

  # 20. Zero Variance
  data <- generate_data(n = 10, kind = "constant") # all 5.0
  run_scenario("20_zero_variance", data$x, data$y, frac = 0.5, deg = 1, iter = 0)

  cat("\nAll scenarios completed successfully!\n")
}

# Run main function
main()
